keps/sig-node/20191205-container-streaming-requests.md

```diff
@@ -0,0 +1,211 @@
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+There are several features and flags for handling container streaming requests
+(exec/attach/port-forward). Having so many different ways of handling these requests unnecessarily
+complicates configuration, and opens the cluster to potential security risks. This is a proposal to
+consolidate the current options into a single supported configuration. Specifically, it proposes:
+
+1. Deprecate the Kubelet's `--redirect-container-streaming` feature
+2. Deprecate the `StreamingProxyRedirects` feature
+
+## Motivation
+
+The `StreamingProxyRedirects` feature was originally added as part of the migration the [Container
+Runtime Interface][]. One of the initial goals was to bypass the Kubelet for these long-running
+requests in order to better account the request overhead. However, the apiserver is still
+responsible for proxying these requests, so the gains of bypassing the Kubelet are
+marginal. Furthermore, the redirect handling significantly complicates the apiserver stream
+handling, and opens it up to [potential SSRF attacks][], so we added the `ValidateProxyRedirects`
+feature to mitigate those.
+
+When we went to add mTLS support to the runtime streaming server, we realized getting the right
+certs established between the apiserver and the runtime was complicated, but the apiserver already
```

Comment:
Is there more detail on this?  Having the justification be strong due to a list of challenges makes this easier to accept that we just can't fix exec on kubelet.

Reply:
Done. https://github.com/kubernetes/kubernetes/issues/36666#issuecomment-378440458

Reply:
To add a bit more context - fixing (securing) the redirect approach is definitely possible, but requires a lot more work than what is outlined in this KEP, and no one has stepped up to do that work. The current state is an SSRF liability, so I'd like to rip it out sooner rather than later.

If someone wants to re-architect streaming to eliminate all proxies (kubelet & apiserver), I'm all for it, but I (nor anyone on my team) has the time to take that on right now. My recommendation is to start by prototyping out of tree (using ephemeral containers, as suggested), and then consider upstreaming the changes if needed.